Fix NAR tests on Linux+ZFS+normalize

yannham opened the pull request:
Closes #11595.

## Motivation & context

A test added recently checks that when trying to deserialize a NAR with two files that Unicode-normalize to the same result either succeeds on Linux, or fails with an "already exists" error on Darwin. However, failing with an "already exists" error can in fact also happen on Linux, when using ZFS with the proper utf8 and Unicode normalization options set.

This commit fixes the issue by not assuming the behavior from the current system, but just by blindly checking that either one of the two aforementioned possibilities happen, whether on Darwin or on Linux.

This has been tested on a Linux+ZFS+normalize runner here: https://github.com/tweag/nickel/actions/runs/11073351502/job/30769680803?pr=2037

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

Gerg-L reviewed (approved):
works on my machine with normalization set to `formD` lgtm

puckipedia commented:
There's a bit of an issue with this actually; UTF-8 normalisation means that nix builds done on this system may inherently end up with a different NAR hash than those done on a system without normalisation. This may also present as nar files from `cache.nixos` not having the same nar hash.

tomberek commented:
> There's a bit of an issue with this actually; UTF-8 normalisation means that nix builds done on this system may inherently end up with a different NAR hash than those done on a system without normalisation. This may also present as nar files from `cache.nixos` not having the same nar hash.

Is that an issue with this test? or with NAR deserialization?

Mic92 commented:
Search for the `use-case-hack` option, that is used as a mitigation.

roberth commented:
To summarize the underlying issue:

- You may need a similar hack to the case hack in order to use such a store transparently as a vessel for deployments to stores that do not have this normalization.
- This is also an insidious [build impurity](https://hydra.nixos.org/build/274162798/download/1/manual/glossary.html#gloss-purity)  (and those are a fact of life)
  
For now, let's make the tests pass.

We could gather related ideas in
- #11555

roberth commented on tests/functional/nars.sh:
```diff
@@ -88,15 +88,30 @@ touch "$TEST_ROOT/case/xt_CONNMARK.h~nix~case~hack~3"
+# `utf8only` enabled and `normalization` set to anything else than `none`).
+# The deserialization should succeed on most Linux.
+#
+# We test that it either succeeds or fails with "already exists" error.
```
Could we determine the properties of the `$TEST_ROOT`'s file system by creating these semi-duplicate files using other means? That way can at least still test that the Nix behavior matches another program's behavior, and for example catch a bad change that causes Nix to normalize on its own accord on a file system where it shouldn't.

Something like
```bash
touch $TEST_ROOT/poll-unicode-â
touch $TEST_ROOT/poll-unicode-â
n=$(wc -l $(ls $TEST_ROOT | grep poll-unicode-))
normalizing=$((n == 1))
```

yannham replied:
I imagine the second `touch` would error out, in the same way as `nix-store --restore`. But we can still test that `touch` and `nix-store --restore` have the same result, at least :+1:

yannham commented:
I implemented @roberth 's suggestion in the latest force push, tested again on a Linux+ZFS+utf8 normalization runner. I also changed the tests of the return code in the `if`-`then`-`else` from `[[ $code == n ]]` to `(( code == n ))`, which I believe is more appropriate.

yannham commented:
Dear maintainers, is there anything else to be done here? It's probably not a very important change but I fear that if it doesn't move forward it's going to be forgotten for ever :sweat_smile: pinging @roberth once again (sorry if inappropriate)

Gerg-L reviewed (approved):
Works on my system with formD

roberth reviewed (approved):
I might have written the conditional on `touchFilesCount` instead, but it'd be equivalent. :+1:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-10-07-nix-team-meeting-minutes-184/54046/1
